torch/csrc/jit/codegen/cuda/kernel_ir.cpp

```diff
@@ -475,30 +475,56 @@ Allocate::Allocate(
     Passkey passkey,
     Val* buffer,
     MemoryType memory_type,
-    Val* size,
+    std::vector<Val*> sizes,
     bool zero_init)
     : Expr(passkey),
       buffer_(buffer),
       memory_type_(memory_type),
-      size_(size),
+      sizes_(std::move(sizes)),
       zero_init_(zero_init) {
-  if (size_ != nullptr) {
-    TORCH_INTERNAL_ASSERT(size_->isOneInt() || buffer_->isA<TensorView>());
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  if (!sizes_.empty()) {
+    TORCH_INTERNAL_ASSERT(
+        (sizes_.size() == 1 && sizes_[0]->isOneInt()) ||
+        buffer_->isA<TensorView>());
   } else {
     TORCH_INTERNAL_ASSERT(buffer_->isA<TensorView>());
     TORCH_INTERNAL_ASSERT(
         buffer_->as<TensorView>()->memoryType() == memory_type_);
-    kir::IrBuilder ir_builder(GpuLower::current()->kernel());
     const auto domain = buffer_->as<TensorView>()->domain();
-    size_ = domain->nDims() == 0 ? ir_builder.create<Int>(1)
-                                 : domain->axis(0)->extent();
-    for (size_t i = 1; i < domain->nDims(); i++) {
-      size_ = ir_builder.mulExpr(size_, domain->axis(i)->extent());
+    for (auto axis : domain->noReductions()) {
+      sizes_.push_back(axis->extent());
```

Comment:
dumb question.

IIUC, the part the handles the inference in the old `inferAndAlloc` function
```
  const auto domain = tv->domain();
  const auto maybe_rfactor_domain =
      domain->hasRFactor() ? domain->rfactorDomain() : domain->rootDomain();

  for (const auto id : maybe_rfactor_domain) {
    if (id->isReduction() ||
        id->iterType() == IterType::BroadcastWithoutStride) {
      continue;
    }
    const auto inferred_val = expr_eval.evaluate(id->rawExtent());
```
now resides inside `analyzeGlobalAllocationSizes` called via `createAllocExpr`.
So this code in the constructor is not used in this PR but for the TensorShift PR to handle the halo region that mentioned earlier?

Reply:
That version of `inferAndAlloc` still exists and is used for output tensors.

This part of the constructor is also used as it is before. Specifically, it is used for allocating sync buffers for reductions. I changed to use `noReductions` as I believe that's what should be correct, but otherwise I tried not to change any semantics.

Reply:
I renamed that `inferAndAlloc` to `inferAndAllocOutput` for clarification.